File: Binary tree level order transversal.py
# TC - O(n)
# SC - O(n)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root == None:
            return []
        output = []
        queue = deque()
        queue.append(root)
        while (queue):
            sz = len(queue)
            innerlist = []
            for i in range(sz):
                ele = queue.popleft()
                innerlist.append(ele.val)
                if ele.left:
                    queue.append(ele.left)
                if ele.right:
                    queue.append(ele.right)
            output.append(innerlist)
        return output

File: BinaryTreeLevelOrderTransversal.cpp
//Time Complexity O(n)
// Space Complexity O(height of tree)(Due to recursion)
//Problem successfully executed on leetcode
//No problems faced while coading this.



#include<iostream>
#include<vector>
using namespace std;

  struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
  };
 
class Solution {
public:
    vector<vector<int> > result;
    vector<vector<int> > levelOrder(TreeNode* root) {
        if (root==NULL) return result;
        
        dfs(root,0);
        return result;
    }
    
    void dfs(TreeNode* root, int level)
    {
        if(root==NULL)
        {
            return;
        }
        if(result.size()==level)
        {
            vector<int> newList;
            result.push_back(newList);
        }
        result[level].push_back(root->val);
        dfs(root->left,level+1);
        dfs(root->right,level+1);
    }
};